File: backend/filter.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

type Filter struct {
	ignore  []*regexp.Regexp
	include []*regexp.Regexp
}

func defaultFilter() (*Filter, error) {
	return newFilter(
		[]string{"node_modules/", ".git/", "dist/", "build/", ".DS_Store", ".env", ".*ignore"},
		nil,
	)
}

func newFilter(ignorePatterns, includePatterns []string) (*Filter, error) {
	ignore, err := compilePatterns(ignorePatterns)
	if err != nil {
		return nil, err
	}
	include, err := compilePatterns(includePatterns)
	if err != nil {
		return nil, err
	}
	return &Filter{ignore: ignore, include: include}, nil
}

func (f *Filter) shouldIgnore(path string) bool {
	if len(f.include) > 0 && !matchesAny(f.include, path) {
		return true
	}
	return matchesAny(f.ignore, path)
}

func (f *Filter) shouldPruneDirectory(path string) bool {
	return matchesAny(f.ignore, path) || matchesAny(f.ignore, path+"/placeholder")
}

func matchesAny(patterns []*regexp.Regexp, value string) bool {
	for _, pattern := range patterns {
		if pattern.MatchString(value) {
			return true
		}
	}
	return false
}

func compilePatterns(patterns []string) ([]*regexp.Regexp, error) {
	compiled := make([]*regexp.Regexp, 0, len(patterns))
	for _, raw := range patterns {
		pattern := strings.TrimSpace(raw)
		if pattern == "" {
			continue
		}
		normalized := strings.TrimSuffix(pattern, "/")
		if strings.HasSuffix(pattern, "/") || (!strings.Contains(normalized, "/") && !strings.Contains(normalized, ".")) {
			normalized += "/**"
		}
		expression, err := globRegexp(normalized)
		if err != nil {
			return nil, &APIError{Status: 400, Message: fmt.Sprintf("invalid filter rule %q: %v", pattern, err)}
		}
		compiled = append(compiled, expression)
	}
	return compiled, nil
}

func globRegexp(pattern string) (*regexp.Regexp, error) {
	var builder strings.Builder
	builder.WriteString("^")
	for index := 0; index < len(pattern); index++ {
		switch pattern[index] {
		case '*':
			builder.WriteString(".*")
		case '?':
			builder.WriteByte('.')
		case '[':
			end := strings.IndexByte(pattern[index+1:], ']')
			if end < 0 {
				return nil, fmt.Errorf("unterminated character class")
			}
			end += index + 1
			class := pattern[index : end+1]
			if _, err := regexp.Compile("^(?:" + class + ")$"); err != nil {
				return nil, fmt.Errorf("invalid character class")
			}
			builder.WriteString(class)
			index = end
		default:
			builder.WriteString(regexp.QuoteMeta(string(pattern[index])))
		}
	}
	builder.WriteString("$")
	return regexp.Compile(builder.String())
}

func splitPatterns(value string) []string {
	parts := strings.Split(value, ",")
	result := make([]string, 0, len(parts))
	for _, part := range parts {
		if value := strings.TrimSpace(part); value != "" {
			result = append(result, value)
		}
	}
	return result
}

File: backend/src/filter.rs
use globset::{Glob, GlobSet, GlobSetBuilder};

use crate::error::AppError;

#[derive(Clone, Debug)]
pub struct Filter {
    ignore: GlobSet,
    include: Option<GlobSet>,
}

impl Filter {
    pub fn new(ignore_patterns: &[String], include_patterns: &[String]) -> Result<Self, AppError> {
        let ignore = build_set(ignore_patterns)?;
        let include = if include_patterns.is_empty() {
            None
        } else {
            Some(build_set(include_patterns)?)
        };
        Ok(Self { ignore, include })
    }

    pub fn defaults() -> Result<Self, AppError> {
        Self::new(
            &[
                "node_modules/**".into(),
                ".git/**".into(),
                "dist/**".into(),
                "build/**".into(),
                ".DS_Store".into(),
                ".env".into(),
                ".*ignore".into(),
            ],
            &[],
        )
    }

    pub fn should_ignore(&self, path: &str) -> bool {
        self.include.as_ref().is_some_and(|set| !set.is_match(path)) || self.ignore.is_match(path)
    }

    pub fn should_prune_directory(&self, path: &str) -> bool {
        self.ignore.is_match(path) || self.ignore.is_match(format!("{path}/placeholder"))
    }
}

fn build_set(patterns: &[String]) -> Result<GlobSet, AppError> {
    let mut builder = GlobSetBuilder::new();
    for pattern in patterns {
        let normalized = pattern.trim().trim_end_matches('/');
        if normalized.is_empty() {
            continue;
        }
        let glob =
            if pattern.ends_with('/') || !normalized.contains('/') && !normalized.contains('.') {
                format!("{normalized}/**")
            } else {
                normalized.to_string()
            };
        builder
            .add(Glob::new(&glob).map_err(|e| AppError::InvalidFilter(format!("{pattern}: {e}")))?);
    }
    builder
        .build()
        .map_err(|e| AppError::InvalidFilter(e.to_string()))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_globs_and_keeps_invalid_rules_visible() {
        let filter = Filter::new(&["node_modules/".into(), "*.log".into()], &[]).unwrap();
        assert!(filter.should_ignore("node_modules/package.json"));
        assert!(filter.should_ignore("logs/app.log"));
        assert!(!filter.should_ignore("src/main.rs"));
        assert!(Filter::new(&["[".into()], &[]).is_err());
    }
}
